src/factories/flowRunArtifacts.ts

```diff
@@ -0,0 +1,233 @@
+import { BoundsContainer } from '@/models/boundsContainer'
+import { waitForApplication, waitForViewport } from '@/objects'
+import { waitForConfig } from '@/objects/config'
+import { emitter } from '@/objects/events'
+import { layout, waitForSettings } from '@/objects/settings'
+
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export async function flowRunArtifactsFactory() {
+  const application = await waitForApplication()
+  const viewport = await waitForViewport()
+  const config = await waitForConfig()
+  const settings = await waitForSettings()
+
+  const artifacts: Map<string, ArtifactFactory> = new Map()
+  const clusterNodes: ArtifactClusterFactory[] = []
+
+  let container: Container | null = null
+  let internalData: Artifact[] | null = null
+  let nonTemporalAlignmentEngaged = false
+
+  emitter.on('viewportMoved', () => update())
+
+  async function render(newData?: Artifact[]): Promise<void> {
+    if (container) {
+      container.visible = !settings.disableArtifacts
+    }
+
+    if (settings.disableArtifacts) {
+      return
+    }
+
+    if (newData) {
+      internalData = newData
+    }
+
+    if (!internalData) {
+      return
+    }
+
+    if (!container) {
+      createContainer()
+    }
+
+    const promises: Promise<BoundsContainer>[] = []
+
+    for (const artifact of internalData) {
+      promises.push(createArtifact(artifact))
+    }
+
+    await Promise.all(promises)
+  }
+
+  async function createArtifact(artifact: Artifact): Promise<BoundsContainer> {
+    if (artifacts.has(artifact.id)) {
+      return artifacts.get(artifact.id)!.render()
+    }
+
+    const factory = await flowRunArtifactFactory({ type: 'artifact', artifact }) as ArtifactFactory
+
+    artifacts.set(artifact.id, factory)
+
+    container!.addChild(factory.element)
+
+    return factory.render()
+  }
+
+  function createContainer(): void {
+    container = new Container()
+    container.zIndex = DEFAULT_ROOT_ARTIFACT_Z_INDEX
+    application.stage.addChild(container)
+  }
+
+  function update(): void {
+    if (!container || settings.disableArtifacts) {
+      return
+    }
+
+    if (!layout.isTemporal()) {
+      if (!nonTemporalAlignmentEngaged) {
+        clearClusters()
+        alignNonTemporal()
+        nonTemporalAlignmentEngaged = true
+      }
+
+      container.position.x = viewport.scale._x + viewport.worldTransform.tx
+      return
+    }
+
+    nonTemporalAlignmentEngaged = false
+    container.position.x = 0
+    checkLayout()
+  }
+
+  const checkLayout = throttle(async () => {
+    const visibleItems = [...artifacts.values()]
+    const availableClusterNodes = [...clusterNodes]
+    visibleItems.sort((itemA, itemB) => itemA.element.x - itemB.element.x)
+
+    await checkCollisions(visibleItems, availableClusterNodes)
+
+    for (const cluster of availableClusterNodes) {
+      cluster.render()
+    }
+  }, DEFAULT_ROOT_ARTIFACT_COLLISION_THROTTLE)
+
+  async function checkCollisions(
```

Comment:
at some point I'd still like to see if we can abstract this clustering logic out into a separate agnostic service or utility. Perhaps when you get to events it'll make sense to do that.